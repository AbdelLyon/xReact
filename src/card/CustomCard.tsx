import { ForwardedRef, forwardRef, HTMLProps, PropsWithChildren } from "react";

import { cn } from "@/utils";
import {
  Card as CardRoot,
  CardFooter,
  CardTitle,
  CardContent,
} from "@/ui/card";

// Props for the CustomCard component
type CustomCardProps = PropsWithChildren<{
  className?: string;
  width?: number | string;
  height?: number | string;
  clickable?: boolean;
}>;

// Props for the CardImage component
interface CardImageProps extends HTMLProps<HTMLDivElement> {
  height?: string | number;
  width?: string | number;
}

type Props = PropsWithChildren<{
  title: string;
  image: React.ReactNode;
  footer?: React.ReactNode;
  clickable?: boolean;
  heightImage?: number;
  width?: number;
  children: React.ReactNode;
  className?: string;
  classNames?: {
    image?: string;
    title?: string;
    content?: string;
    footer?: string;
  };
}>;

// CardImage component with forwardRef
export const CardImage = forwardRef<HTMLDivElement, CardImageProps>(
  (
    { height = 100, width = 150, className, ...props },
    ref: ForwardedRef<HTMLDivElement>,
  ) => {
    return (
      <div
        ref={ref}
        className={cn(
          "relative overflow-hidden transition-transform duration-300 ease-in-out transform",
          className,
        )}
        data-testid="card-image"
        style={{ height, width }}
        {...props}
      />
    );
  },
);

CardImage.displayName = "CardImage";

export const CustomCard = ({
  className,
  width,
  height,
  clickable = false,
  children,
}: CustomCardProps) => {
  return (
    <CardRoot
      className={cn(
        "relative flex flex-col border rounded-sm transition-transform duration-300 ease-in-out",
        {
          "cursor-pointer hover:shadow-lg": clickable,
        },
        className,
      )}
      data-testid="custom-card"
      style={{ width, height }}
    >
      {children}
    </CardRoot>
  );
};

const Card = ({
  title,
  image,
  className,
  footer,
  clickable = false,
  heightImage = 200,
  width = 300,
  classNames,
  children,
}: Props) => {
  return (
    <CustomCard
      className={`overflow-hidden ${className}`}
      clickable={clickable}
      width={width}
    >
      <CardImage
        className={`rounded-t-md ${classNames?.image}`}
        height={heightImage}
        width={width}
      >
        {image}
      </CardImage>
      <CardTitle className={`mt-2 px-4 ${classNames?.title}`}>
        <h2 className="text-lg font-semibold">{title}</h2>
      </CardTitle>
      <CardContent className={`mt-2 px-4 ${classNames?.content}`}>
        {children}
      </CardContent>
      {footer && (
        <CardFooter
          className={`flex w-full justify-end px-4 py-2 ${classNames?.footer}`}
        >
          {footer}
        </CardFooter>
      )}
    </CustomCard>
  );
};

export { Card };
