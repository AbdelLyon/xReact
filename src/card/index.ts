import "@/index.css";
export { Card } from "@/card/CustomCard";
