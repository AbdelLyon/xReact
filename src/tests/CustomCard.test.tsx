import { render } from "@testing-library/react";
import { describe, expect, test } from "vitest";

import { CardImage, CustomCard } from "@/card/CustomCard";

describe("CardImage", () => {
  test("should match snapshot when rendering children", () => {
    const { asFragment } = render(<CardImage>Test Child</CardImage>);

    expect(asFragment()).toMatchSnapshot();
  });

  test("should match snapshot with custom classes", () => {
    const { asFragment } = render(
      <CardImage className="custom-class">Children</CardImage>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("should match snapshot when forwarded ref is used", () => {
    const ref = { current: null };
    const { asFragment } = render(<CardImage ref={ref}>Children</CardImage>);

    expect(asFragment()).toMatchSnapshot();
  });
});

describe("CustomCard", () => {
  test("should match snapshot when rendering children", () => {
    const { asFragment } = render(<CustomCard>Test Child</CustomCard>);

    expect(asFragment()).toMatchSnapshot();
  });

  test("should match snapshot with clickable styles", () => {
    const { asFragment } = render(<CustomCard clickable />);

    expect(asFragment()).toMatchSnapshot();
  });

  test("should match snapshot with width and height styles", () => {
    const { asFragment } = render(<CustomCard height="100px" width="200px" />);

    expect(asFragment()).toMatchSnapshot();
  });
});
