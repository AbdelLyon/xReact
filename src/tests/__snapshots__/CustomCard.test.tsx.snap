// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardImage > should match snapshot when forwarded ref is used 1`] = `
<DocumentFragment>
  <div
    class="relative overflow-hidden transition-transform duration-300 ease-in-out transform"
    data-testid="card-image"
    style="height: 100px; width: 150px;"
  >
    Children
  </div>
</DocumentFragment>
`;

exports[`CardImage > should match snapshot when rendering children 1`] = `
<DocumentFragment>
  <div
    class="relative overflow-hidden transition-transform duration-300 ease-in-out transform"
    data-testid="card-image"
    style="height: 100px; width: 150px;"
  >
    Test Child
  </div>
</DocumentFragment>
`;

exports[`CardImage > should match snapshot with custom classes 1`] = `
<DocumentFragment>
  <div
    class="relative overflow-hidden transition-transform duration-300 ease-in-out transform custom-class"
    data-testid="card-image"
    style="height: 100px; width: 150px;"
  >
    Children
  </div>
</DocumentFragment>
`;

exports[`CustomCard > should match snapshot when rendering children 1`] = `
<DocumentFragment>
  <div
    class="border-border bg-white dark:bg-background text-card-foreground shadow-sm relative flex flex-col border rounded-sm transition-transform duration-300 ease-in-out"
    data-testid="custom-card"
  >
    Test Child
  </div>
</DocumentFragment>
`;

exports[`CustomCard > should match snapshot with clickable styles 1`] = `
<DocumentFragment>
  <div
    class="border-border bg-white dark:bg-background text-card-foreground shadow-sm relative flex flex-col border rounded-sm transition-transform duration-300 ease-in-out cursor-pointer hover:shadow-lg"
    data-testid="custom-card"
  />
</DocumentFragment>
`;

exports[`CustomCard > should match snapshot with width and height styles 1`] = `
<DocumentFragment>
  <div
    class="border-border bg-white dark:bg-background text-card-foreground shadow-sm relative flex flex-col border rounded-sm transition-transform duration-300 ease-in-out"
    data-testid="custom-card"
    style="width: 200px; height: 100px;"
  />
</DocumentFragment>
`;
